### PR TITLE
4849 partner logo

### PR DESCRIPTION
**Pull request by Jodge**
# Description

Branding webapp - ability to customize web title, favicon, and logo

medic/medic-webapp#4849

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/src/routing.js [hunk 1]
```diff
@@ -95,7 +96,17 @@ if (process.argv.slice(2).includes('--allow-cors')) {
...
+    pouch.medic.get('branding').then(doc => {
+      pouch.medic.getAttachment(doc._id, doc.resources.favicon).then(blob => {
+        res.send(blob);
+      });
+    }).catch(err => {
+      res.end(err);
+    });
+  } else {
+    next();
+  }
```
Comment (garethbowen) [inline api/src/routing.js]: This is not the right place to be handling this. When you see `app.use` like this it's middleware for doing things like checking every request for auth or whatever. What you want is a standard route handler, eg:

```js
app.get('/favicon.ico', function(req, res) {
  ...
});
```

These start around line 210 in the file.

You will probably have to remove the _other_ favicon from the `public` dir otherwise it'll take precedence.

Also we don't need two favicons so `git mv` the one from `api/src/public/favicon.ico` to the new location at `api/src/resources/ico/favicon.ico`.

File: api/src/routing.js [hunk 2]
```diff
@@ -95,7 +96,17 @@ if (process.argv.slice(2).includes('--allow-cors')) {
...
 app.use((req, res, next) => {
   req.id = uuid.v4();
-  next();
+  if (req.originalUrl === '/favicon.ico') {
+    pouch.medic.get('branding').then(doc => {
+      pouch.medic.getAttachment(doc._id, doc.resources.favicon).then(blob => {
+        res.send(blob);
+      });
+    }).catch(err => {
+      res.end(err);
```
Comment (garethbowen) [inline api/src/routing.js]: Use `serverUtils.serverError` to finish the response nicely.

File: webapp/src/css/enketo/_fonts.scss [hunk 3]
```diff
@@ -1,8 +1,8 @@
 @font-face {
     font-family: 'enketo-icons';
-    src: url('../fonts/enketo-icons-v2.woff') format('woff'),
-        url('../fonts/enketo-icons-v2.ttf') format('truetype'),
-        url('../fonts/enketo-icons-v2.svg#enketo-icons') format('svg');
+    src: url('/fonts/enketo-icons-v2.woff') format('woff'),
+        url('/fonts/enketo-icons-v2.ttf') format('truetype'),
+        url('/fonts/enketo-icons-v2.svg#enketo-icons') format('svg');
```
Comment (garethbowen) [inline webapp/src/css/enketo/_fonts.scss]: Why are these fonts moving? It seems unrelated to the branding change?

File: admin/src/js/filters/header-logo.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+angular.module('filters').filter('headerLogo',
+  function($sce, BrandingImages) {
+    'use strict';
+    'ngInject';
+    return function(name) {
+      return $sce.trustAsHtml(BrandingImages.getLogo(name));
+    };
+  }
+);
```
[the remark this answers is not in the document]
Comment (Jodge) [reply admin/src/js/filters/header-logo.js]: I thought so but the problem I had was that the filters for webapp is registering the name `inboxFilters` for **angular.module** while in admin it is `filters`

File: webapp/src/css/enketo/_fonts.scss (hunk 3 above)
Comment (Jodge) [reply webapp/src/css/enketo/_fonts.scss]: This shouldn't be committed, a separate task should be created for this. This came up when I realized that the admin app could not get `NotoSans` font, solution was to create `fonts` folder on the admin app and add `NotoSans-Regular.tff` to it. The only problem with that would be duplication since they already existed on the webapp. So I  decided to serve all the fonts from a central location i.e. `api public` dir

Comment (garethbowen) [reply webapp/src/css/enketo/_fonts.scss]: If you serve them from there they probably won't be cached for offline use... but we can discuss that in a separate commit!

File: api/src/migrations/add-branding-doc.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  }).then(doc => {
+    const srcLogo = path.join(__dirname, '..', `resources/${logo.dir}`, logo.file);
+    const srcFav = path.join(__dirname, '..', `resources/${favicon.dir}`, favicon.file);
+    attachDocument(srcLogo, doc, logo).then(doc => {
+      attachDocument(srcFav, doc, favicon);
+    });
+  })
+  .catch(err => {
+    logger.error('%o',err);
+  });
```
Comment (garethbowen) [inline api/src/migrations/add-branding-doc.js]: We need to throw this error to stop API from starting up if this migration fails. Just delete these 3 lines so it bubbles out to the migration runner.

File: api/src/migrations/add-branding-doc.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    resources: {
+      [logo.name]: logo.file,
+      [favicon.name]: favicon.file
+    }
+  }).then(doc => {
+    const srcLogo = path.join(__dirname, '..', `resources/${logo.dir}`, logo.file);
+    const srcFav = path.join(__dirname, '..', `resources/${favicon.dir}`, favicon.file);
+    attachDocument(srcLogo, doc, logo).then(doc => {
+      attachDocument(srcFav, doc, favicon);
+    });
```
Comment (garethbowen) [inline api/src/migrations/add-branding-doc.js]: Attachments can be added to the doc before it's put so it all gets saved in one `put`. This will save a bit of time and reduce the number of sequences in the couchdb changes feed so it's good practice. We do this in webapp in the AddAttachment service - just copy that code into API and use that instead of using `putAttachment`.

File: api/src/migrations/add-branding-doc.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  });
+};
+
+module.exports = {
+  name: 'add-branding-doc',
+  created: new Date(2018, 11, 22, 10, 0, 0, 0),
+  run: () => {
+    return createDoc();
+  }
+};
```
Comment (garethbowen) [inline api/src/migrations/add-branding-doc.js]: Nice one merging these three migrations!

File: api/src/routing.js [hunk 8]
```diff
@@ -151,6 +153,16 @@ app.get('/', function(req, res) {
   }
 });
 
+app.get('/favicon.ico', (req, res) => {
+  db.medic.get('branding').then(doc => {
+    db.medic.getAttachment(doc._id, doc.resources.favicon).then(blob => {
```
Comment (garethbowen) [inline api/src/routing.js]: The name is hardcoded here, right? So we shouldn't need to get the `branding` doc at all. eg:

```js
db.medic.getAttachment('branding', 'favicon').then(blob => {
```

File: admin/src/templates/images_branding.html [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        <div ng-bind-html="logo | headerLogo"></div>
+      </div>
+    </div>
+  </div>
+</div>
+<div class="row">
+  <form id="favicon-upload">
+    <legend translate>upload.favicon</legend>
+    <div class="form-group required">
+      <label translate>admin.app.favicon </label> <span ng-bind-html="favicon"></span>
```
Comment (garethbowen) [inline admin/src/templates/images_branding.html]: Last one - how about `branding.favicon.field`?

File: api/src/routing.js (hunk 8 above)
Comment (Jodge) [reply api/src/routing.js]: Hmm attachment are named based on file name so `favicon` above should be `favicon.ico` so if a partner decide to  upload a favicon with the name `lg.ico` we will have to get that name from the `doc.resources` which keeps track of the attachments. That is why I am fetching the doc and getting the name of the  attachment `doc.resources.favicon` i.e. this is key value based

Comment (garethbowen) [reply api/src/routing.js]: Ahh, that makes sense. Good call.
